Resources/doc/reference/security.rst

```diff
@@ -74,89 +74,97 @@ Let's create the following strategy : a media can be downloaded only once per se
-         * @param \Symfony\Component\DependencyInjection\ContainerInterface $container
-         * @param int $times
+         * @param TranslatorInterface $translator
+         * @param Session             $session
+         * @param int                 $times
          */
-        public function __construct(TranslatorInterface $translator, ContainerInterface $container, $times)
+        public function __construct(TranslatorInterface $translator, Session $session, $times)
         {
-            $this->times    = $times;
-            $this->container = $container;
+            $this->times = $times;
+            $this->session = $session;
             $this->translator = $translator;
         }
 
         /**
-         * @param \Sonata\MediaBundle\Model\MediaInterface $media
-         * @param \Symfony\Component\HttpFoundation\Request $request
-         * @return bool
+         * {@inheritdoc}
          */
         public function isGranted(MediaInterface $media, Request $request)
         {
-            if (!$this->container->has('session')) {
-                return false;
-            }
-
-            $times = $this->getSession()->get($this->sessionKey, 0);
+            $times = $this->session->get($this->sessionKey, 0);
 
             if ($times >= $this->times) {
                 return false;
             }
 
-            $this->getSession()->set($this->sessionKey, $times + 1);
+            $this->session->set($this->sessionKey, $times + 1);
 
             return true;
         }
 
         /**
-         * @return string
+         * {@inheritdoc}
          */
         public function getDescription()
         {
             return $this->translator->trans('description.session_download_strategy', array('%times%' => $this->times), 'SonataMediaBundle');
         }
-
-        /**
-         * @return \Symfony\Component\HttpFoundation\Session
-         */
-        private function getSession()
-        {
-            return $this->container->get('session');
-        }
     }
 
 Let's explain a bit :
 
-* ``__construct`` : the constructor get the different dependency. As the session belongs to the request scope, it is not possible to inject the service session, so the Container is injected.
 * ``isGranted`` : the method test the number of time the file has been downloaded
 * ``getDescription`` : return a translated message to explain what the current strategy does
-* ``getSession`` : return the session from the container.
 
 
-The last important part is declaring the service. Open the ``service.xml`` file and add the following lines.
+The last important part is declaring the service.
+
+Open the ``service.xml`` file and add the following lines.
 
 .. code-block:: xml
 
-        <service id="sonata.media.security.session_strategy" class="Sonata\MediaBundle\Security\SessionDownloadStrategy" >
+        <service id="sonata.media.security.session_strategy" class="Sonata\MediaBundle\Security\SessionDownloadStrategy" scope="request">
```

Comment:
Not sure I undestand what you mean. What I'm saying is that we should detect support for shared services and use the right service definition depending on that. Building the service definition should be done in https://github.com/sonata-project/SonataMediaBundle/blob/3.x/DependencyInjection/SonataMediaExtension.php